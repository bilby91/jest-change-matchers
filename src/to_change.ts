
export async function toChange(
  receivedMutator: () => Promise<any>,
  target: () => Promise<any>,
) {
  const previousValue = await target()
  await receivedMutator()
  const newValue = await target()

  const pass = previousValue !== newValue
  const message = pass
    ? () =>
        `Expected target not to change.\n` +
        `Received:\n` +
        `  ${newValue}\n`
    : () =>
        `Expected target to change.\n` +
        `Received:\n` +
        `  ${newValue}\n`

  return { message, pass }
}
