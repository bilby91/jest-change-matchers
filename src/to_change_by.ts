
export async function toChangeBy(
  receivedMutator: () => Promise<number>,
  target: () => Promise<number>,
  changeCount: number,
) {
  const previousLength = await target()
  await receivedMutator()
  const newLength = await target()
  const changeDifference = newLength - previousLength

  const pass = changeDifference === changeCount
  const message = pass
    ? () =>
        `Expected target not to have changed by:\n` +
        `  ${changeCount}\n` +
        `Received:\n` +
        `  ${changeDifference}\n`
    : () =>
        `Expected target to have changed by:\n` +
        `  ${changeCount}\n` +
        `Received:\n` +
        `  ${changeDifference}\n`

  return { message, pass }
}
