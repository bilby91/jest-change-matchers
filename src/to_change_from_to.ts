export async function toChangeFromTo(
  receivedMutator: () => Promise<any>,
  target: () => Promise<any>,
  from: any,
  to: any,
) {
  const previousValue = await target()
  await receivedMutator()
  const newValue = await target()

  const pass = previousValue === from && newValue === to
  const message = pass
    ? () =>
        `Expected target not to have changed\n` +
        `  from: ${from}\n` +
        `  to: ${to}\n` +
        `Received:\n` +
        `  from: ${previousValue}\n` +
        `  to: ${newValue}\n`
    : () =>
        `Expected target to have changed\n` +
        `  from: ${from}\n` +
        `  to: ${to}\n` +
        `Received:\n` +
        `  from: ${previousValue}\n` +
        `  to: ${newValue}\n`

  return { message, pass }
}
