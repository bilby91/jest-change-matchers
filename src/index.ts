import "jest"

/// <reference path="./matchers.d.ts" />

import { toChange } from "./to_change"
import { toChangeBy } from "./to_change_by"
import { toChangeFromTo } from "./to_change_from_to"

expect.extend({
  toChange: toChange as any,
  toChangeBy: toChangeBy as any,
  toChangeFromTo: toChangeFromTo as any,
})
