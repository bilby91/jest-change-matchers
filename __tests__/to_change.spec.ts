import "jest"
import "../src"

let sharedState = 0

describe("toChange", () => {
  beforeEach(() => {
    sharedState = 0
  })

  describe(".toChange", () => {
    it("passes when there is a change", async () => {
      await expect(async () => {
        sharedState += 1
      }).toChange(async () => {
        return sharedState
      })
    })

    it("throws an error when there is no change", async () => {
      const testCase = () => {
        return expect(async () => {
          sharedState += 0
        }).toChange(async () => {
          return sharedState
        })
      }

      await expect(testCase()).rejects.toThrowErrorMatchingSnapshot()
    })
  })

  describe(".not.toChange", () => {
    it("passes when there is no change", async () => {
      await expect(async () => {
        // no-op
      }).not.toChange(async () => {
        return sharedState
      })
    })

    it("throws an error when there is a change", async () => {
      const testCase = () => {
        return expect(async () => {
          sharedState += 1
        }).not.toChange(async () => {
          return sharedState
        })
      }

      await expect(testCase()).rejects.toThrowErrorMatchingSnapshot()
    })
  })
})
