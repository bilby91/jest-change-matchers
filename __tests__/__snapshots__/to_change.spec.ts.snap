// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toChange .not.toChange throws an error when there is a change 1`] = `
"Expected target not to change.
Received:
  1
"
`;

exports[`toChange .toChange throws an error when there is no change 1`] = `
"Expected target to change.
Received:
  0
"
`;
